Source/scrollrt.cpp

```diff
@@ -1293,7 +1337,41 @@ static void DrawMain(int dwHgt, BOOL draw_desc, BOOL draw_hp, BOOL draw_mana, BO
 		if (sgdwCursWdt != 0) {
 			DoBlitScreen(sgdwCursX, sgdwCursY, sgdwCursWdt, sgdwCursHgt);
 		}
-	}
+	}*/
+
+/*instead blitting everything on surface with slight decrease of performance*/
+    DoBlitScreen(0, 0, 640, 480); 
+
+/* scaling operation start */
+extern SDL_Surface *renderer_texture_surface; /** defined in dx.cpp */
+extern SDL_Surface *scale2x_renderer_texture_surface; /** surface for scaling defined in dx.cpp */
+
+    unsigned int *pixels = (unsigned int*)renderer_texture_surface->pixels;
+	unsigned int *scale2x_pixels = (unsigned int*)scale2x_renderer_texture_surface->pixels;
+
+
+    for(int y=0; y<480;y++)
+    {
+        unsigned int *src0, *src1, *src2, *dst0, *dst1;
+
+        dst0 = &scale2x_pixels[2*y*640*2];
+        dst1 = &scale2x_pixels[(2*y+1)*640*2];
+
+        int y_prev=y-1 , y_next=y+1;
+        if(y_prev < 0) y_prev=0;
+        if(y_next > 480-1) y_next=480-1;
+
+        src0 = &pixels[y_prev*640*2];
+        src1 = &pixels[y*640*2];
+        src2 = &pixels[y_next*640*2];
+
+        scale2x_32_def_single(dst0, src0, src1, src2, 640);

```

Comment:
Probably just use SDL_BlitScaled() instead of creating your own function that operates on individual pixels.